core/genesis/file.go

```diff
@@ -0,0 +1,311 @@
+package genesis
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"google.golang.org/protobuf/proto"
+
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+type FileGenesisSource struct {
+	sourceDir           string
+	moduleName          string
+	moduleRootJson      json.RawMessage // the RawMessage from the genesis.json app_state.<module> that got passed into InitCHain
+	readFromRootGenesis bool
+}
+
+const (
+	fileOpenflag  = os.O_CREATE | os.O_WRONLY
+	flieOpenMode  = fs.FileMode(0o600)
+	dirCreateMode = fs.FileMode(0o700)
+)
+
+// readCloserWrapper is io.ReadCloser wrapper for bytes.Reader
+type readCloserWrapper struct {
+	io.Reader
+}
+
+// Close implements io.Closer interface
+func (r readCloserWrapper) Close() error {
+	return nil
+}
+
+// NewFileGenesisSource returns a new GenesisSource for the provided
+// source directory and the provided module name where it is assumed
+// that the encoded json data of file.
+func NewFileGenesisSource(sourceDir, moduleName string) GenesisSource {
+	return &FileGenesisSource{sourceDir: filepath.Clean(sourceDir), moduleName: moduleName}
+}
+
+// OpenReader opens the source field reading from the given parameters,
+// and returns a ReadCloser.
+// It will try to open the field in order following by:
+// <sourceDir>/<module>/<field>.json
+// <field> key inside <sourceDir>/<module>.json
+// app_state.<module>.<field> key in <sourceDir>/genesis.json
+func (f *FileGenesisSource) OpenReader(field string) (io.ReadCloser, error) {
+	var rawBz json.RawMessage
+
+	// if moduleRootJson is not nil, we can skip reading data from file
+	if f.moduleRootJson != nil {
+		rawBz = f.moduleRootJson
+	}
+
+	if rawBz == nil {
+		// try reading genesis data from <sourceDir>/<module>/<field>.json
+		fName := fmt.Sprintf("%s.json", field)
+		fPath := filepath.Join(f.sourceDir, f.moduleName)
+
+		fp, err := os.Open(filepath.Clean(filepath.Join(fPath, fName)))
+		if err == nil {
+			return fp, nil
+		}
+
+		// if cannot find it, try reading from <sourceDir>/<module>.json
```

Comment:
The file reading logic was suggested in https://github.com/cosmos/cosmos-sdk/issues/12808#issuecomment-1233082028. So it will return not found error until checked `module.json` and then `genesis.json`

Reply:
I see.